### DISCUSSION TITLE
Need GPIO pull down for Wyze Plug Outdoor

### DISCUSSION THREAD

**Discussion by Frogmore42** [arendst/Tasmota#10814] [Ideas]
So I bought a couple of the new Wyze Plug Outdoor devices.  They have an ESP32 in them and it is possible to flash Tasmota on them.

They have a BL0937 power measuring IC which is working okay.  The base voltage was quite away off, but easy to calibrate.  Power numbers looked right on, as did current.

They have independent LEDs from the relays and I have that working, along with the status/link LED.

But, the buttons have a 10K pull-up to one side and the GPIO on the other side.  There appears to be some other conditioning circuitry and it looks like they are counting on the ESP32 to pull down the pin.  It looks like the ESP32 has that capability, but it doesn't seem like Tasmota supports that currently. 

This is the template so far:
```
{"NAME":"Wyze Plug Outdoor","GPIO":[1,1,1,1,1,576,1,1,1,1,1,224,321,1,1,320,0,1,1,1,0,2624,2656,2720,0,0,0,0,225,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}
```

**Comment by neel-m** [comment]
As a test, I modified my local copy of Tasmota code to use Pull Down instead of Pull Up for buttons to see if I could get the buttons to work on the Wyze Plug Outdoor.  That solved the problem for me.  It looks like it will be a more involved change to make it selectable.  I took a quick look at the code and it looks like there is no quick/easy way to add this capability.  Is there any documentation on the thinking for GPIO specification?  I didn't see any, but didn't look too hard either.  It seems like it might be generally useful to support other devices that choose to design input circuitry that needs it.

**Reply by barbudor** [reply]
So we already have:
Button : active low, internal pull-up
Button_n : active low, no internal pull-up
Button_i : active high with internal pull-up
Button_ni: active high with internal pull-up

So on ESP32 we could have a
Button_d : active low, with internal pull-down
Button_di : active high, with internal pull-down
This would need to create new components codes in https://tasmota.github.io/docs/Components/#gpio-conversion
with the proper text for the pull down menu, etc ....
Looks feasible but would requires preferably Theo to assign the new numbers

Not sure what you are looking about "GPIO specification", may be the table linked above already provide what you are looking for

**Reply by Frogmore42** [reply]
So, can @arendst assign numbers for these?
I can create a PR to implement them, or I suppose I could just assign the next available numbers.

**Reply by arendst** [reply]
Just noticed the request.

I will work on it. If someone could just assure me this feature would never pop-up on an ESP8266....

**Reply by barbudor** [reply]
AFAIK, ESP8266 do not offer pull down on GPIO 0..15, only pull-up. GPIO 16 only has pull-down, no pull-up

**Reply by arendst** [reply]
OK. I'll (try to) implement it for ESP32 today.

**Comment by arendst** [comment]
Implemented in latest dev vesion as ``Button_d`` and ``Button_id``

**Comment by davlaw** [comment]
Since these changes have been implemented is there an update template?

**Reply by Frogmore42** [reply]
The changes to allow the required pull-down are in place.  I have not tried them yet, so have not updated the template.  I'll see if I can get to it in the near future.

**Comment by davlaw** [comment]
I think digiblur commented in his video that the buttons were odd, each button was connected to a different resistor value to a gpio pin.

**Comment by DWTSD** [comment]
Hello,
New to Tasmota, so still trying to figure things out.  Just bought two of the Wyze Outdoor Plugs.  After a slight struggle I was able to flash with the current version of Tasmota.  It took a while and several flashings, I realized that the USB to Serial adapter doesn't provide enough current for the chip to boot.  (see first attached photo)
Now that I have flashed and calibrated both, I have one problem and one question.  The problem is that button one doesn't cycle relay one.  Button two does cycle relay two.  I can cycle both via the web interface.
I've noticed that GPIO17 has a value of Button_id 2, but none of the GPIO's have a value of Button_id 1.  Is that what is missing?

My question is:  There is a label Analog 1 and in the screen shot it has a value of 1309.  This value changes but I cannot find a correlation with any commands.  Does anyone know what this is?

THIA
dale
![ULNF2fsI6YXCYJUM](https://github.com/arendst/Tasmota/assets/152427993/e8da3b17-0602-4602-b859-41c01f60037f)

**Reply by sfromis** [reply]
Assuming that you mean this one: https://templates.blakadder.com/wyze_WLPPO1.html - you should not do the config manually by reading the reference table to the right, but instead follow the "How to use Templates?" link in the left column to apply the template, in one simple copy/paste operation.

The thing is that the formatting of the reference table is broken, when gpio18 shows up as "74x595 OE" instead of the correct "Button_id 1", which is indeed what was missing in the translation there, but it is correctly included when copy/pasting the template. Maybe @blakadder can fix this, to make the pin show up the correct way.

Looks like the analog input on gpio 34 is the Lux sensor on top of the device. It might make sense to config as `ADC Light` instead of the "raw" `ADC Input`. Then you can config the reading using the command [AdcParam](https://tasmota.github.io/docs/ADC/#commands), number 3 for for the Lux equation parameters, if the defaults are not correct.

**Reply by DWTSD** [reply]
Thanks sfromis,
You are correct, I entered it from the table, then copied from from someone's "how to flash" web page which was incorrect.  When I copied from the blakadder.com page it is correct.  I also switched gpio34 to ADC Light and it makes much more sense!  Thanks for all of your help!
Dale
![image](https://github.com/arendst/Tasmota/assets/152427993/23a79a69-4d3e-4f58-bcc2-8107be10c618)
